llvm/lib/Target/AMDGPU/AMDGPUTargetMachine.cpp

```diff
@@ -951,6 +951,10 @@ bool AMDGPUTargetMachine::isNoopAddrSpaceCast(unsigned SrcAS,
 }
 
 unsigned AMDGPUTargetMachine::getAssumedAddrSpace(const Value *V) const {
+  if (auto *Arg = dyn_cast<Argument>(V);
+      Arg && AMDGPU::isKernelCC(Arg->getParent()))
+    return AMDGPUAS::GLOBAL_ADDRESS;
```

Comment:
Probably need to defend against byref values. Ideally we would only use byref, in which case this needs to look at the load from the argument. But then if we're changing the IR calling convention lowering we might as well fix it to emit addrspace(1) in the first place

Reply:
If we have something like `ptr byref(i32) %in.byref`, I suppose it is also in AS1?

Reply:
I don't know if we are able to do device memory allocation in AS4 but I rule out `byref` argument anyway.

> But if we're fixing clang to do the right thing, shouldn't need to handle the flat case

We will do that, but the fix in clang will take more time and I really want to get rid of the special handling in the attributor.